tripal/t4d8
# Tv4g1 129 tripal fields complex

# Issues Addressed
Issue #259
Issue #258 (partially)

# Description
This PR includes a first draft organism field.  This PR makes some important changes to how fields are mapped to Chado. Currently the support for fields is for single value fields and the mapping of fields to Chado was stored in the field settings.  The problem for complex fields is that not all properties for a field will be "stored" some are simply "loaded".  For example, all of the content types derived from the feature type will have an organism.  The organism field only needs to "store" the `organism_id` value.  But for display we need more information as part of the field. These value are "loaded" from the foreign key with the organism table.  

Additionally, some fields need their values set after loading data from Chado. An example of this is the label field for the organism. It combines the values for the genus, species and infraspecific type/name.  

To address this problem. we only specify the Chado `base_table` for a field. All fields are associated with a single base record and the `base_table` indicates which table the `record_id` belongs to.   Each property of a field is then responsible for mapping itself to a table and column of Chado in relation to the base table.  

For single value fields here is what the storage properties look like:
```php
'storage_settings' => [
   'storage_plugin_id' => 'chado_storage',
      'storage_plugin_settings' => [
         'base_table' => $chado_table,
         'property_settings' => [
            'value' => [
               'action' => 'store',
               'chado_table' => $chado_table,
               'chado_column' => $column,
            ],
         ],
      ],
   ],
],
```
You can see the `base_table` is specified, but there is also a new `property_settings` key.  Inside this array, any property that needs customization can be set.  The key is the property key.  The `action` indicates what to do with this property's value.  The `action` of `store` simply tells the storage backend the value gets stored in the specified chado table and column.  

For a complex field such as the organism field, we still have something similar when creating the field:  

```php
'storage_settings' => [
     'storage_plugin_id' => 'chado_storage',
     'storage_plugin_settings' => [
        'base_table' => $chado_table,
         'property_settings' => [
            'value' => [
            'action' => 'store',
            'chado_table' => $chado_table,
            'chado_column' => $org_id_col,
         ],
      ],
   ],
],
```
Notice that here the `value` property also has an action of `store`. This is because the `value` property maps to the `organism_id` column of the table (e.g. `feature`).  However,  this field has several other properties including the label, genus, species, infraspecific_name, etc.  The user creating the field doesn't need to customize how those properties are saved, because they will always be saved the same no matter what is the base table.  Therefore, the settings for these properties are set in the `tripalTypes` function of the field class.   For example:

```php
   // Indicate the action to perform for each property.
    $settings = $field_definition->getSetting('storage_plugin_settings');
    $base_table = $settings['base_table'];
    $value_settings = $settings['property_settings']['value'];
    $label_settings = [
      'action' => 'replace',
      // @todo: we should probably change this to a function so that we have
      // better control over how these labels are created. There are functions
      // in Tripal v3 to do this.
      'template' => "<i>[$genus_term] [$species_term]</i> [$iftype_term] [$ifname_term]",
    ];
    $genus_settings = [
      'action' => 'join',
      'path' => $base_table . '.organism_id>organism.organism_id',
      'chado_column' => 'genus'
    ];
    $species_settings = [
      'action' => 'join',
      'path' => $base_table . '.organism_id>organism.organism_id',
      'chado_column' => 'species'
    ];
    $iftype_settings = [
      'action' => 'join',
      'path' => $base_table . '.organism_id>organism.organism_id;organism.type_id>cvterm.cvterm_id',
      'chado_column' => 'name',
      'as' => 'infraspecific_type_name'
    ];
    $ifname_settings = [
      'action' => 'join',
      'path' => $base_table . '.organism_id>organism.organism_id',
      'chado_column' => 'infraspecific_name',
    ];

    // Create the property types.
    $value = new IntStoragePropertyType($entity_type_id, self::$id, 'value', $value_settings);
    $label = new VarCharStoragePropertyType($entity_type_id, self::$id, $label_term, $label_len, $label_settings);
    $genus = new VarCharStoragePropertyType($entity_type_id, self::$id, $genus_term, $genus_len, $genus_settings);
    $species = new VarCharStoragePropertyType($entity_type_id, self::$id, $species_term, $species_len, $species_settings);
    $ifname = new VarCharStoragePropertyType($entity_type_id, self::$id, $ifname_term, $ifname_len, $ifname_settings);
    $iftype = new IntStoragePropertyType($entity_type_id, self::$id, $iftype_term, $iftype_settings);

```
Notice the `value_settings` is pulled from the field configuration and later passed into the value property object.  But the other properties are manually set.  

There are two other actions in the code above:  `join` and `replace`.  The `join` action tells the storage back-end which Chado tables the field needs to join to get its value.  It also indicates the column it will use for it's value and specify the column alias it wants to use if there is a potential conflict on a join (e.g. cvterm.name and feature.name).  The specified `chado_column` is always a column in the table at the end of the series of joins in the path.

The `replace` action replaces the tokens in the string with the values loaded on other fields.  

I do want to add a `function` action so that developers, if need be, can write their own functions to set the values. But that is not included in this PR.   This is needed for generating a proper label for the organism. We have a lot of code that @dsenalik cleaned up in Tripal v3 that we need to use here. 

# What's Left
1.  We need to deal with the issue if the base table and the chado table the field stores to are not in the same table (e.g. the property field).   
2.  We need to decide how to hide fields (and hence store their values) for content types that shouldn't have them. For example, we don't want a cvterm field for the feature table content types, as that's built into the content type.  We need to store and save the `type_id` but not via a visible field.   I propose we add to our alpha1 release resolution of this problem as well as the organism and property fields.

# Testing
1.  Completely uninstall Drupal and or start with a new fresh install.
3. Download this branch and install `tripal`, `tripal_biodb` and `tripal_chado`.
4. Install Chado
5. Prepare Chado
6. Create new organism page
7. Create a new gene page. The page should be created without any errors.  You should notice the following:
   1. When creating the gene page the organism field will have a drop down with the organism you added in step 5.  
   2. When saving the organism the page shows the organism scientific name and not the organism_id
   3. We don't yet have a complex field for the type, and actually the type should not be present at all because this is built into the content type. But for now, you can't save a gene without entering the type_id.  Just use number 159 (for gene).

As long as all went well the field works!
